File: dinner_planner/lib/services/nutrition_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

class NutritionService {

  static const apiKey = "VlQkgbKYMMonAfxW2tsvkDSdiDzFoa4cgyOoUaox";

  static Future<List> searchFood(String query) async {

    final url = Uri.parse(
      "https://api.nal.usda.gov/fdc/v1/foods/search?query=$query&api_key=$apiKey"
    );

    final response = await http.get(url);

    if (response.statusCode != 200) {
      return [];
    }

    final data = jsonDecode(response.body);

    return data["foods"] ?? [];
  }

  static Future<Map<String, dynamic>?> lookupBarcode(String barcode) async {
    final url = Uri.parse(
      'https://world.openfoodfacts.org/api/v0/product/$barcode.json',
    );
    final response = await http.get(url);
    if (response.statusCode != 200) return null;
    final data = jsonDecode(response.body);
    if (data['status'] != 1) return null;
    final product = data['product'] as Map<String, dynamic>;
    final nutriments = (product['nutriments'] as Map<String, dynamic>?) ?? {};
    return {
      'name': (product['product_name'] as String?)?.trim() ?? 'Unknown Product',
      'calories': (nutriments['energy-kcal_100g'] as num?)?.toDouble() ?? 0,
      'protein': (nutriments['proteins_100g'] as num?)?.toDouble() ?? 0,
      'carbs': (nutriments['carbohydrates_100g'] as num?)?.toDouble() ?? 0,
      'fat': (nutriments['fat_100g'] as num?)?.toDouble() ?? 0,
    };
  }

  static Map<String, double> extractNutrition(List nutrients) {

    double calories = 0;
    double protein = 0;
    double carbs = 0;
    double fat = 0;

    for (var n in nutrients) {

      switch (n["nutrientId"]) {

        case 1008:
          calories = (n["value"] ?? 0).toDouble();
          break;

        case 1003:
          protein = (n["value"] ?? 0).toDouble();
          break;

        case 1005:
          carbs = (n["value"] ?? 0).toDouble();
          break;

        case 1004:
          fat = (n["value"] ?? 0).toDouble();
          break;
      }
    }

    return {
      "calories": calories,
      "protein": protein,
      "carbs": carbs,
      "fat": fat
    };
  }
}

File: web/src/services/nutritionService.ts
const USDA_API_KEY = 'VlQkgbKYMMonAfxW2tsvkDSdiDzFoa4cgyOoUaox'

export async function searchFood(query: string): Promise<any[]> {
  const url = `https://api.nal.usda.gov/fdc/v1/foods/search?query=${encodeURIComponent(query)}&api_key=${USDA_API_KEY}`
  const response = await fetch(url)
  if (!response.ok) return []
  const data = await response.json()
  return data.foods ?? []
}

export interface BarcodeProduct {
  name: string
  calories: number
  protein: number
  carbs: number
  fat: number
}

export async function lookupBarcode(barcode: string): Promise<BarcodeProduct | null> {
  const url = `https://world.openfoodfacts.org/api/v0/product/${encodeURIComponent(barcode)}.json`
  const response = await fetch(url)
  if (!response.ok) return null
  const data = await response.json()
  if (data.status !== 1) return null
  const product = data.product ?? {}
  const nutriments = product.nutriments ?? {}
  return {
    name: (product.product_name as string | undefined)?.trim() || 'Unknown Product',
    calories: Number(nutriments['energy-kcal_100g'] ?? 0),
    protein: Number(nutriments['proteins_100g'] ?? 0),
    carbs: Number(nutriments['carbohydrates_100g'] ?? 0),
    fat: Number(nutriments['fat_100g'] ?? 0),
  }
}

export interface ExtractedNutrition {
  calories: number
  protein: number
  carbs: number
  fat: number
}

export function extractNutrition(nutrients: any[]): ExtractedNutrition {
  let calories = 0
  let protein = 0
  let carbs = 0
  let fat = 0

  for (const n of nutrients) {
    switch (n.nutrientId) {
      case 1008:
        calories = n.value ?? 0
        break
      case 1003:
        protein = n.value ?? 0
        break
      case 1005:
        carbs = n.value ?? 0
        break
      case 1004:
        fat = n.value ?? 0
        break
    }
  }

  return { calories, protein, carbs, fat }
}
